webapp/src/Form/Type/UserRegistrationType.php

```diff
@@ -85,6 +87,39 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'mapped' => false,
             ]);
 
+        $registrationCategoryNames = $this->dj->dbconfig_get('registration_category_names', []);
+        $showCategoryChoice        = count($registrationCategoryNames) !== 1;
+        $registrationCategories    = [];
+        foreach ($registrationCategoryNames as $name) {
+            $category = $this->em->getRepository(TeamCategory::class)->findOneBy(['name' => $name]);
```

Comment:
Why not do a single query:
```php
$registrationCategories = $this->em->getRepository(TeamCategory::class)
                             ->findBy(['name' => $registrationCategoryNames]);
```
That's simpler and does less DB queries.

Reply:
It doesn't preserve the order of the categories, which I think is important.
Would [this](https://github.com/tom93/domjudge/compare/self-registration-multiple-categories-v2-conditionally-build-field..tom93:self-registration-multiple-categories-v2b-one-db-query) be better? (And should I do the same in `CheckConfigService.php`?)

```diff
--- a/webapp/src/Form/Type/UserRegistrationType.php
+++ b/webapp/src/Form/Type/UserRegistrationType.php
@@ -89,12 +89,16 @@ class UserRegistrationType extends AbstractType
 
         $registrationCategoryNames = $this->dj->dbconfig_get('registration_category_names', []);
         if (count($registrationCategoryNames) > 1) {
+            $unorderedCategories = $this->em->getRepository(TeamCategory::class)->findBy(['name' => $registrationCategoryNames]);
+            $categoryMap         = [];
+            foreach ($unorderedCategories as $category) {
+                $categoryMap[$category->getName()] = $category;
+            }
             $registrationCategories = [];
             foreach ($registrationCategoryNames as $name) {
-                $category = $this->em->getRepository(TeamCategory::class)->findOneBy(['name' => $name]);
                 // silently ignore invalid category names
-                if ($category) {
-                    $registrationCategories[] = $category;
+                if (array_key_exists($name, $categoryMap)) {
+                    $registrationCategories[] = $categoryMap[$name];
                 }
             }
             // if all the category names are invalid, the user won't be able to fill in the form; don't even show it
```

Reply:
I'd prefer this indeed, although I'd also sacrifice the order to simplify the code.
The code in `CheckConfigService.php` is less important as it gets called rarely; there I'd use whichever is simplest.

Reply:
I changed it to use

    $registrationCategories    = $this->em->getRepository(TeamCategory::class)->findBy(
        ['name' => $registrationCategoryNames],
        ['sortorder' => 'ASC']
    );

which matches the order of the categories in the scoreboard.